src/control/src/point_solver/psolver.cpp

```diff
@@ -64,49 +63,88 @@ std::pair<Position, int> PointSolver::update_closest_point(
  * @return std::pair<Position, int> lookahead point and error status (1 = error)
  */
 std::tuple<Position, double, bool> PointSolver::update_lookahead_point(
-    const std::vector<custom_interfaces::msg::PathPoint> &pathpoint_array, Position rear_axis_point,
+    const std::vector<custom_interfaces::msg::PathPoint> &pathpoint_array,
     int closest_point_id) const {
-  Position lookahead_point = Position();
-
-  double closest_distance = std::numeric_limits<double>::max();
-  custom_interfaces::msg::PathPoint closest_yet;
-
+  Position rear_axis_point = this->vehicle_pose_.rear_axis_;
   double ld = this->k_ * std::max(this->vehicle_pose_.velocity_, 2.0);
   RCLCPP_DEBUG(rclcpp::get_logger("control"), "Current ld: %f", ld);
-  
-  for (size_t i = 0; i < pathpoint_array.size(); i++) {
-    size_t index = (closest_point_id + i) % pathpoint_array.size();
-    lookahead_point = Position(pathpoint_array[index].x, pathpoint_array[index].y);
-    double distance = rear_axis_point.euclidean_distance(lookahead_point);
 
-    // RCLCPP_INFO(rclcpp::get_logger("control"), "Distance to lookahead point %f, index %lu", distance, index);
-
-    // if the distance is within the margin, return the point
-    if (std::abs(distance - ld) <= (ld * this->ld_margin_)) {
-      return std::make_tuple(lookahead_point, pathpoint_array[index].v, false);
+  for (size_t i = 0; i < pathpoint_array.size(); i++) {
+    size_t index_a = (closest_point_id + i) % pathpoint_array.size();
+    size_t index_b = (closest_point_id + i + 1) % pathpoint_array.size();
+    auto point_a = Position(pathpoint_array[index_a].x, pathpoint_array[index_a].y);
+    auto point_b = Position(pathpoint_array[index_b].x, pathpoint_array[index_b].y);
+
+    if (!(rear_axis_point.euclidean_distance(point_a) < ld &&
+          rear_axis_point.euclidean_distance(point_b) > ld)) {
+      continue;
     }
 
-    // if the distance is less than the lookahead distance (we don't want a lookahead further than
-    // ld if it's not withing ld margin), check if it is the closest point found yet
-    if (distance > 1 && std::abs(ld - distance) < closest_distance) {
-      RCLCPP_INFO(rclcpp::get_logger("control"), "New closest point found that doesnt fint margin: %f", distance);
-      closest_distance = std::abs(ld - distance);
-      closest_yet = pathpoint_array[index];
+    double x;
+    double y;
+
+    // Slope of the line is infinite, don't need line equation
+    if (point_a.x == point_b.x) {
+      RCLCPP_DEBUG(rclcpp::get_logger("control"), "Vertical line!!");
+      x = point_a.x;
+      double delta = ld * ld - std::pow(x - rear_axis_point.x, 2);
```

Comment:
Why not use abs instead of squaring the numbers? I am not understanding the purpose of the delta variable

Reply:
n percebi o comentario, o delta e pq e feito sqrt e n queremos fazer isso se o delta for negativo